pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/FloatDictionary.java

```diff
@@ -30,7 +31,24 @@ public FloatDictionary(PinotDataBuffer dataBuffer, int length) {
 
   @Override
   public int insertionIndexOf(String stringValue) {
-    return binarySearch(Float.parseFloat(stringValue));
+    // First convert string to BigDecimal and then downcast to int. This allows type conversion from any compatible
+    // numerical value represented as string to an int value.
+    BigDecimal bigDecimal = new BigDecimal(stringValue);
```

Comment:
This is a FloatDictionary, so we can safely assume that the caller stack ensure it is called with a stringValue that is expected to be float parsed. I don't think we should convert it into BigDecimal.

Reply:
This doesn't appear to be the case. For example if the query is:
    `SELECT count(*) from mytable WHERE intColumn < 3000000000.453`
then `stringValue` gets set to 3000000000.453 and the current code would throw exception because 3000000000.453 is not a valid integer.

After the change, the above query will evaluate successfully (which is expected SQL behavior). Same with other queries such as:
    `SELECT count(*) from mytable WHERE intColumn = 5.0

Reply:
What is the expected behavior from other SQL engines if the value does not match the column data type: `intColumn < 3000000000.453`?

Reply:
The expected behavior is that the query will still evaluate and produce results as long as the numerical value can be upcasted/downcasted to the column type. For example, the following works on MySQL and PostgreSQL.

```
create table test(intColumn int, floatColumn float);
insert into test values(5, 5.5);
insert into test values(10, 15.56);

select * from test where intColumn < 300.453;
select * from test where intColumn < 300000000000000000000000.453;
select * from test where floatColumn > 6;
```
300000000000000000000000.453 is outside the bounds of Integer datatype, so it can't be represented as an Integer value, but gets parsed to something that can be compared to an integer.